web-api/src/storage/postgres.rs

```diff
@@ -382,56 +446,96 @@ impl Database {
                 .await?;
         }
 
-        let mut builder = QueryBuilder::new(
-            "UPDATE document
-            SET is_candidate = TRUE
-            WHERE document_id IN ",
-        );
-        let mut ingested = Vec::with_capacity(ingestable.len());
-        let mut ids = ingestable.iter().peekable();
-        while ids.peek().is_some() {
-            ingested.extend(
-                builder
-                    .reset()
-                    .push_tuple(ids.by_ref().take(Self::BIND_LIMIT))
-                    .push(" RETURNING document_id, snippet, is_summarized, properties, tags, embedding;")
-                    .build()
-                    .try_map(|row| {
-                        let (id, snippet, is_summarized, Json(properties), tags, embedding) =
-                            FromRow::from_row(&row)?;
-                        Ok(IngestedDocument {
-                            id,
-                            snippet,
-                            is_summarized,
-                            properties,
-                            tags,
-                            embedding,
-                            is_candidate: true,
-                        })
-                    })
-                    .fetch_all(&mut tx)
-                    .await?,
-            );
-        }
+        let needs_ingestion =
+            Self::set_is_candidate_and_return_for_ingestion(&mut tx, ingestable.iter().copied())
+                .await?;
 
         tx.commit().await?;
 
-        let failed = (ingested.len() < ingestable.len())
+        let failed = (needs_ingestion.len() < ingestable.len())
             .then(|| {
-                for document in &ingested {
+                for document in &needs_ingestion {
                     ingestable.remove(&document.id);
                 }
                 ingestable.into_iter().cloned().collect()
             })
             .unwrap_or_default();
 
-        Ok((unchanged, ingested, failed))
+        Ok((unchanged, needs_ingestion, failed))
+    }
+
+    async fn set_is_candidate_and_return_for_ingestion(
+        tx: &mut Transaction<'_, Postgres>,
+        ids: impl ExactSizeIterator<Item = &DocumentId> + Clone,
+    ) -> Result<Vec<DocumentForIngestion>, Error> {
+        let mut builder = QueryBuilder::new(
+            "SELECT document_id, snippet_idx, snippet, embedding
+            FROM snippet
+            WHERE document_id IN ",
+        );
+        let mut snippets = HashMap::<_, BTreeMap<_, _>>::new();
```

Comment:
:art: can you add a comment regarding the btreemap? it took me a moment to figure out that we use this to return the snippets of each document ordered by their index.

Reply:
:+1:  just used a `Vec` and `sorted()` which shouldn't be slower